### PR TITLE
Refresh IM attribute encoding for attributeData and attribute report and corresponding read/write logic

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Replace AttributeDataElement with AttributeDataIB and Add AttributeReport and AttributeReports per latest IM encoding spec, and update corresponding IM read/write interaction with attributeReport. 
This PR would cover lots of renaming per latest IM spec.
This PR also replace all FieldId with attributeId in IM

#### Change overview
See above.

#### Testing
Add unit test for attribute report.
Existing test covers.

### CODE REVIEW

File: src/app/MessageDef/AttributeDatas.cpp [hunk 1]
```diff
@@ -17,11 +17,11 @@
  */
 /**
  *    @file
- *      This file defines AttributeDataList parser and builder in CHIP interaction model
+ *      This file defines AttributeDatas parser and builder in CHIP interaction model
```
Comment (yufengwangca) [inline src/app/MessageDef/AttributeDatas.cpp:20]: nit: 'Data' is already a plural of datum, AttributeDatas -> AttributeData

File: src/app/MessageDef/AttributeDatas.cpp [hunk 2]
```diff
@@ -37,13 +37,13 @@ using namespace chip::TLV;
 namespace chip {
 namespace app {
 #if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
-CHIP_ERROR AttributeDataList::Parser::CheckSchemaValidity() const
+CHIP_ERROR AttributeDatas::Parser::CheckSchemaValidity() const
```
Comment (yufengwangca) [inline src/app/MessageDef/AttributeDatas.cpp:40]: nit: 'Data' is already a plural of datum, AttributeDatas -> AttributeData

**Review by yufengwangca — APPROVED** [review]

File: src/app/MessageDef/AttributeDatas.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef/AttributeDatas.cpp]: Done, rename to AttribtueDataIBs. thanks

File: src/app/MessageDef/AttributeDatas.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/MessageDef/AttributeDatas.cpp]: Done, rename to AttribtueDataIBs. thanks

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/ClusterInfo.h [hunk 3]
```diff
@@ -48,11 +48,11 @@ struct ClusterInfo
     static constexpr ListIndex kInvalidListIndex = 0xFFFF;
 
 public:
-    bool IsAttributePathSupersetOf(const ClusterInfo & other) const
+    bool IsAttributePathIBsupersetOf(const ClusterInfo & other) const
```
Comment (mrjerryjohns) [inline src/app/ClusterInfo.h:51]: I'm not so certain this name change is needed. ClusterInfo != AttributePathIB, so naming it like this is not quite accurate.

File: src/app/InteractionModelEngine.h [hunk 4]
```diff
@@ -255,26 +255,18 @@ void DispatchSingleClusterResponseCommand(const ConcreteCommandPath & aCommandPa
 bool ServerClusterCommandExists(const ConcreteCommandPath & aCommandPath);
 
 /**
- *  Fetch attribute value and version info and write to the TLVWriter provided.
- *  When the endpoint / cluster / attribute / event data specified by aClusterInfo does not exist, corresponding interaction model
- * error code will be put into the writer, and CHIP_NO_ERROR will be returned and apDataExists will be set to false.
- *  If the data exists on the server, the data (with tag kCsTag_Data) and the data version (with tag kCsTag_DataVersion) will be put
- * into the TLVWriter and apDataExists will be set to true. TLVWriter error will be returned if any error occurred during encoding
- * these values.
+ *  Fetch attribute value and version info and write to the AttributeReport provided.
```
Comment (mrjerryjohns) [inline src/app/InteractionModelEngine.h:258]: The existing comments around error generation are still valid, but perhaps, need to get re-worded in the context of `AttributeReportIB`.

**Review by jelderton — APPROVED** [review]

File: src/app/ClusterInfo.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/ClusterInfo.h:51]: agree, just revert this renaming.

File: src/app/InteractionModelEngine.h (hunk 4 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:258]: agree, just reword regarding error part, thanks

File: src/app/ReadClient.cpp [hunk 5]
```diff
@@ -442,78 +441,79 @@ void ReadClient::OnResponseTimeout(Messaging::ExchangeContext * apExchangeContex
...
+        err = report.GetAttributeStatus(&status);
+        if (CHIP_NO_ERROR == err)
         {
-            // The spec requires that one of data or status code must exist, thus failure to read data and status code means we
-            // received malformed data from server.
-            SuccessOrExit(err = element.GetStatus(&statusU16));
-            status = static_cast<Protocols::InteractionModel::Status>(statusU16);
+            StatusIB::Parser errorStatus;
+            ReturnErrorOnFailure(status.GetPath(&path));
+            ReturnErrorOnFailure(ProcessAttributePath(path, clusterInfo));
```
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:494]: Followup: `ProcessAttributePath` should just directly write into a ConcreteAttributePath, without the ClusterInfo temporary.

File: src/app/util/ember-compatibility-functions.cpp [hunk 6]
```diff
@@ -187,189 +187,215 @@ bool ServerClusterCommandExists(const ConcreteCommandPath & aCommandPath)
...
 
     AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aPath.mEndpointId, aPath.mClusterId);
     if (attrOverride != nullptr)
     {
         // TODO: We should probably clone the writer and convert failures here
         // into status responses, unless our caller already does that.
-        AttributeValueEncoder valueEncoder(apWriter, aAccessingFabricIndex);
+        writer = attributeDataIBBuilder.GetWriter();
+        VerifyOrReturnError(writer != nullptr, CHIP_NO_ERROR);
+        AttributeValueEncoder valueEncoder(writer, aAccessingFabricIndex);
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:218]: If we now plan to guarantee that `AttributeValueEncoder` has a writer, we can simplify its code to get rid of null-checks...  Do we plan to guarantee that?

File: src/lib/core/CHIPError.cpp [hunk 7]
```diff
@@ -665,6 +665,9 @@ bool FormatCHIPError(char * buf, uint16_t bufSize, CHIP_ERROR err)
     case CHIP_ERROR_IM_MALFORMED_INVOKE_RESPONSE_MESSAGE.AsInteger():
         desc = "Malformed Interaction Model Invoke Response MESSAGE";
         break;
+    case CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_REPORT_MESSAGE.AsInteger():
+        desc = "Malformed Interaction Model Attribute Report MESSAGE";
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.cpp:669]: Why the all-caps "MESSAGE"?

**Pull request merged**